parameter/criterion/src/Criterion.cpp

```diff
@@ -0,0 +1,296 @@
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "criterion/Criterion.h"
+#include <Utility.h>
+#include <log/Logger.h>
+
+#include <stdexcept>
+#include <sstream>
+#include <cassert>
+
+namespace core
+{
+namespace selection
+{
+namespace criterion
+{
+
+Criterion::Criterion(const std::string& name, const Values& values, core::log::Logger& logger)
+    : Criterion(name, logger, values, State{values.begin()->second}, {})
+{
+    if (mValues.size() < 2) {
+        throw InvalidCriterionError("Not enough values were provided for exclusive criterion '" +
+                                    mName  + "' which needs at least 2 values");
+    }
+}
+
+Criterion::Criterion(const std::string& name,
+                     core::log::Logger& logger,
+                     const Values& values,
+                     const State&& defaultState,
+                     const MatchMethods& derivedMatchMethods)
+    : mValues(values),
+      mMatchMethods(gatherMap<MatchMethods>(MatchMethods{
+                                {"Is", [&](const State& state){ return mState == state; }},
+                                {"IsNot", [&](const State& state){ return mState != state; }}},
+                                derivedMatchMethods)),
+      mState(), _uiNbModifications(0), _logger(logger), mName(name), mDefaultState(defaultState)
+{
+    setDefaultState();
+}
+
+bool Criterion::hasBeenModified() const
+{
+    return _uiNbModifications != 0;
+}
+
+void Criterion::resetModifiedStatus()
+{
+    _uiNbModifications = 0;
+}
+
+bool Criterion::setDefaultState()
+{
+    if (mState == mDefaultState) {
+        // Default state is already set, nothing to do
+        return false;
+    }
+    mState = mDefaultState;
+    return true;
+}
+
+bool Criterion::setState(const State& state, std::string& error)
+{
+    if (state.size() > 1) {
+        error = "Exclusive criterion '" + mName + "' can't be set with more than one value";
+        return false;
+    }
+
+    // Check for a change
+    if (mState != state) {
+        bool stateChanged = false;
+        if (state.empty()) {
+            stateChanged = setDefaultState();
```

Comment:
That's not an issue: regardless of what value we are setting, we can check whether it is the same as the current one.

    State oldState = mState;
    if (state.empty()) {
        state = mDefaultState;
    } else {
        if (!getLiteralValue(*state.begin(), error)) {
            error = ...;
            return false;
        }
    }

    mState = state;

    if (mState != oldState) {
        stateModificationsEvent();
    }

    return true;